### ISSUE TITLE
PathsAndConsumesOfModules can return wrong modules in conjunction with EDAlias

### ISSUE THREAD

**Issue by makortel** [cms-sw/cmssw#30937]
The following configuration
```python
import FWCore.ParameterSet.Config as cms

process = cms.Process("TEST")

process.maxEvents.input = 1
process.source = cms.Source("EmptySource")

intEventProducer = cms.EDProducer("IntProducer", ivalue = cms.int32(1))

intEventConsumer = cms.EDAnalyzer("IntTestAnalyzer",
    moduleLabel = cms.untracked.string(""),
    valueMustMatch = cms.untracked.int32(1)
)

process.intProducerFoo = intEventProducer.clone(ivalue = 1)
process.intProducerBar = intEventProducer.clone(ivalue = 2)
process.intConsumer = intEventConsumer.clone(moduleLabel = "intProducer")

process.intProducer = cms.EDAlias(
    intProducerFoo = cms.VPSet(
        cms.PSet(
            type = cms.string("edmtestIntProduct"),
        )
    ),
    intProducerBar = cms.VPSet(
        cms.PSet(
            type = cms.string("edmtestIntProduct"),
            fromProductInstance = cms.string(""),
            toProductInstance = cms.string("test"),
        )
    )
)

process.t = cms.Task(
    process.intProducerFoo,
    process.intProducerBar
)
process.p = cms.Path(
    process.intConsumer,
    process.t
)

process.Tracer = cms.Service("Tracer", dumpPathsAndConsumes = cms.untracked.bool(True))
``` 
leads to following printout from Tracer service
```
All modules and modules in the current process whose products they consume:
...
  IntTestAnalyzer/'intConsumer' consumes products from these modules:
    IntProducer/'intProducerBar'
```
while the `intProducer` is really consuming the product from `intProducerFoo`.

**Comment by makortel** [comment]
assign core

**Comment by makortel** [comment]
I think the problem is in `EDConsumerBase::modulesWhoseProductsAreConsumed()`, more specifically how EDAliases are handled in
https://github.com/cms-sw/cmssw/blob/8d03f366426f05d38169387b6404ab070a7a3ad4/FWCore/Framework/src/EDConsumerBase.cc#L445-L458
where the alias-to module label is translated to alias-from module label **only by module labels**.

The problematic EDAlias, on the other hand, has one alias-to label but two alias-from module labels, so the logic above picks the the one earlier in the ordered vector.

It seems to me that logic, together with
https://github.com/cms-sw/cmssw/blob/8d03f366426f05d38169387b6404ab070a7a3ad4/DataFormats/Provenance/interface/ProductRegistry.h#L121-L123
need to be changed to follow the aliases on branch/product basis instead of by module labels alone.

**Comment by makortel** [comment]
Fixed in #31512.
